internal/api/oidc/auth_request.go

```diff
@@ -593,19 +576,140 @@ func CreateCodeCallbackURL(ctx context.Context, authReq op.AuthRequest, authoriz
 	return callback, err
 }
 
-func CreateTokenCallbackURL(ctx context.Context, req op.AuthRequest, authorizer op.Authorizer) (string, error) {
-	client, err := authorizer.Storage().GetClientByClientID(ctx, req.GetClientID())
+func (s *Server) CreateTokenCallbackURL(ctx context.Context, req op.AuthRequest) (string, error) {
+	provider := s.Provider()
+	opClient, err := provider.Storage().GetClientByClientID(ctx, req.GetClientID())
 	if err != nil {
 		return "", err
 	}
-	createAccessToken := req.GetResponseType() != oidc.ResponseTypeIDTokenOnly
-	resp, err := op.CreateTokenResponse(ctx, req, client, authorizer, createAccessToken, "", "")
+	client, ok := opClient.(*Client)
+	if !ok {
+		return "", zerrors.ThrowInternal(nil, "OIDC-waeN6", "Error.Internal")
+	}
+
+	session, state, err := s.command.CreateOIDCSessionFromAuthRequest(
+		setContextUserSystem(ctx),
+		req.GetID(),
+		implicitFlowComplianceChecker(),
+	)
 	if err != nil {
 		return "", err
 	}
-	callback, err := op.AuthResponseURL(req.GetRedirectURI(), req.GetResponseType(), req.GetResponseMode(), resp, authorizer.Encoder())
+	resp, err := s.accessTokenResponseFromSession(ctx, client, session, state, client.client.ProjectID, client.client.ProjectRoleAssertion)
+	if err != nil {
+		return "", err
+	}
+	callback, err := op.AuthResponseURL(req.GetRedirectURI(), req.GetResponseType(), req.GetResponseMode(), resp, provider.Encoder())
 	if err != nil {
 		return "", err
 	}
 	return callback, err
 }
+
+func implicitFlowComplianceChecker() command.AuthRequestComplianceChecker {
+	return func(_ context.Context, authReq *command.AuthRequestWriteModel) error {
+		if err := authReq.CheckAuthenticated(); err != nil {
+			return err
+		}
+		return nil
+	}
+}
+
+func (s *Server) authorizeCallbackHandler(w http.ResponseWriter, r *http.Request) {
+	authorizer := s.Provider()
+
+	func() (err error) {
+		ctx, span := tracing.NewSpan(r.Context())
+		r = r.WithContext(ctx)
+		defer func() { span.EndWithError(err) }()
+
+		id, err := op.ParseAuthorizeCallbackRequest(r)
+		if err != nil {
+			op.AuthRequestError(w, r, nil, err, authorizer)
+			return err
+		}
+		authReq, err := authorizer.Storage().AuthRequestByID(r.Context(), id)
+		if err != nil {
+			op.AuthRequestError(w, r, nil, err, authorizer)
+			return err
+		}
+		if !authReq.Done() {
+			op.AuthRequestError(w, r, authReq,
+				oidc.ErrInteractionRequired().WithDescription("Unfortunately, the user may be not logged in and/or additional interaction is required."),
+				authorizer)
+			return err
+		}
+		return s.authResponse(authReq, authorizer, w, r)
```

Comment:
do we even need the error return?

Reply:
I do the error return so the `defer func() { span.EndWithError(err) }()` works properly. AFAIK, if I would declare a `err` variable in the function body, it would remain `nil` if an error was returned from a function where the declaraction and assignment operator (`v, err := ...`) is used.